src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -0,0 +1,1024 @@
+    out = torch.nn.functional.dropout(x + bias, p=prob, training=training)
+    out = residual + out
+    return out
+
+
+def get_bias_dropout_add(training):
+    def _bias_dropout_add(x, bias, residual, prob):
+        return bias_dropout_add(x, bias, residual, prob, training)
+
+    return _bias_dropout_add
+
+
+@torch.jit.script
+def bias_dropout_add_fused_train(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, True)
+
+
+@torch.jit.script
+def bias_dropout_add_fused_inference(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, False)
+
+
+class BLOOMAttention(nn.Module):
+    def __init__(self, config, layer_number=None):
+        super().__init__()
+
+        dtype = getattr(torch, config.dtype)
+        self.pretraining_tp = config.pretraining_tp
+
+        self.hidden_size = config.hidden_size
+        self.num_heads = config.n_head
+        self.head_dim = self.hidden_size // self.num_heads
+        self.split_size = self.hidden_size
+        self.attention_softmax_in_fp32 = config.attention_softmax_in_fp32
+        self.masked_softmax_fusion = config.masked_softmax_fusion
+
+        if dtype == torch.float16:
+            self.fp16 = True
+            self.bf16 = False
+        elif dtype == torch.bfloat16:
+            self.fp16 = False
+            self.bf16 = True
+        else:
+            self.fp16 = False
+            self.bf16 = False
+
+        if self.head_dim * self.num_heads != self.hidden_size:
+            raise ValueError(
+                f"`hidden_size` must be divisible by num_heads (got `hidden_size`: {self.hidden_size} and `num_heads`: {self.num_heads})."
+            )
+
+        # Layer-wise attention scaling
+        self.layer_number = max(1, layer_number)
+        coeff = self.layer_number
+        self.norm_factor = math.sqrt(self.head_dim) * coeff
+
+        # self.scale_mask_softmax = nn.Softmax(dim=1)
+        self.scale_mask_softmax = ScaledSoftmax(
+            self.fp16,
+            self.bf16,
+            self.masked_softmax_fusion,
+            attention_mask_func,
+            self.attention_softmax_in_fp32,
+            coeff,
+            config.seq_length,
+        )
+
+        self.query_key_value = nn.Linear(self.hidden_size, 3 * self.hidden_size, dtype=dtype, bias=True)
+        # TODO : Try a custom class
+        self.dense = nn.Linear(self.hidden_size, self.hidden_size, dtype=dtype)
+        self.skip_bias_add = config.skip_bias_add
+        self.skip_bias_add_qkv = config.skip_bias_add_qkv
+        self.attention_dropout = torch.nn.Dropout(config.attention_dropout)
+
+    def forward(
+        self,
+        hidden_states,
+        layer_past=None,
+        attention_mask=None,
+        alibi=None,
+        head_mask=None,
+        use_cache=False,
+        output_attentions=False,
+    ):
+        # hidden_states: [sq, b, h]
```

Comment:
The comment would be much more helpful with full names for the variables :-)
(here and everywhere below)